### PR TITLE
User configurable R location

### PR DESCRIPTION
**Pull request by jennybc**
Addresses #2235 

Adds 2 configuration points:

* `positron.r.customRootFolders` is an array. Each entry is a folder that is searched for 1 or more R installations.
* `positron.r.customBinaries` is an array. Each entry is the path to a specific R binary.

### QA Notes

It's hard to get into a position where this configuration comes into play, because you have to make sure that four methods of automatic discovery _won't work_. I'm focusing on Windows because that's the main OS that drives the need for this.

* Install R at a non-standard location. This will have to be via the CRAN installer since rig doesn't give you any choice about this. Therefore I removed my rig-installed released version of R and re-installed released R with the CRAN installer below `C:/nonstandardRLocation/`. Make sure to **uncheck the box** to write to the registry! If you forget or you're doing this with an existing installation, you can remove the registry keys with `C:/nonstandardRLocation/R-4.4.1/bin/x64/RSetReg.exe /U` (note the `/U` argument for **un**setting keys). All R installations on Windows ship with this executable for setting/unsetting registry keys.
* Now use rig (or other means) to make *some other R version* the current version. In my setup, R 4.2.3 and 4.3.3 are good candidates for the new current version of R, e.g. `rig default 4.3.3`.
* Make sure this R installation is not on the PATH.
* Fire up Positron. The interpreter drop down should not include this R version that is
  - not in the standard root folder for R, i.e. not in `C:/Program Files/R` nor in `C:/Users/<Username>/AppData/Local/Programs/R`
  - not the current version of R
  - not recorded in the registry
  - not on the PATH
 
Use the command palette to do *Preferences: Open Settings*. Go to *Extensions > R > Positron R Language Pack (advanced)* and do each of these, in turn:

* In *Custom Binaries*, add the path to the nonstandard R binary, e.g. `C:/nonstandardRLocation/R-4.4.1/bin/x64/R.exe`
* In *Custom Root Folders*, add the parent folder of the nonstandard R version, e.g. `C:/nonstandardRLocation`

Fire up Positron again. Either of these settings should make the nonstandard installation appear in the interpreter dropdown.

Other things that can be noticed in the Positron R output channel:

* In the presence of custom binaries, you'll see something like this:
  
      2024-10-30 01:03:07.761 [info] User-specified R binaries:
      [
        "C:/nonstandardRLocation/R-4.4.1-bin/x64.R.exe"
      ]
* In the presence of a custom root folder, you'll see something like this:

      2024-10-30 00:56:15.137 [info] User-specified directories to scan for R installations:
      [
        "C:/nonstandardRLocation"
      ]

### CODE REVIEW

File: extensions/positron-r/src/provider.ts [hunk 1]
```diff
@@ -33,10 +33,11 @@ export const R_DOCUMENT_SELECTORS = [
 enum BinarySource {
 	/* eslint-disable-next-line @typescript-eslint/naming-convention */
 	HQ = 'HQ',
-	adHoc = 'ad hoc locations',
+	adHoc = 'ad hoc location',
 	registry = 'Windows registry',
 	/* eslint-disable-next-line @typescript-eslint/naming-convention */
-	PATH = 'PATH'
+	PATH = 'PATH',
+	user = 'user-specified directory'
```
Comment (jennybc) [inline extensions/positron-r/src/provider.ts:40]: Just continuing to lay groundwork that will be helpful when tackling #1397.

File: extensions/positron-r/src/provider.ts [hunk 2]
```diff
@@ -401,22 +441,34 @@ export async function findCurrentRBinary(): Promise<string | undefined> {
...
 
+	// TODO: for macOS, this should arguably be whatever
+	// /Library/Frameworks/R.framework/Versions/Current/ resolves to
+	// that would remove overlap between `findCurrentBinary()` and `findRBinaryFromPATH()`
+
 	cachedRBinary = await findRBinaryFromPATH();
 	return cachedRBinary;
 }
 
+let cachedRBinaryFromPATH: string | undefined;
```
Comment (jennybc) [inline extensions/positron-r/src/provider.ts:452]: This is an unrelated improvement. Part of continuing effort to reduce rediscovery and repetition in the output channel.

**Review by juliasilge — APPROVED** [review]
From a code perspective and how the options are presented to users, this is looking great!

I did just spend a bit of time trying to install R in a non-standard location on my dev machine AND FAILED 😆 so let's get this merged in and available for testing in a broader set of circumstances.
